pydsa/linked_list.py

```diff
@@ -0,0 +1,127 @@
+class LinkedListNode(object):
+    """
+        Node of a Singly Linked List.
+
+        Contains a value and a pointer to the next node. Both params are optional
+
+        >>> second = LinkedListNode(2)
+        >>> first = LinkedListNode(1, second)
+        >>> first.value
+        1
+        >>> first.get_next()
+        [2]
+
+    """
+    def __init__(self, value=None, next_node=None):
+        self.value = value
+        self.next_node = next_node
+
+    def get_next(self):
+        return self.next_node
+
+    def get_value(self):
+        return self.value
+
+    def set_next(self, next_node):
+        self.next_node = next_node
+
+    def __str__(self):
+        return "[%s]"%(str(self.value))
+
+    def __repr__(self):
+        return str(self)
+
+class LinkedList(object):
+    """
+    Singly Linked List.
+
+    list.head is the head node of the list.
+
+    >>> list = LinkedList()
+
+    Methods implemented:
+
+    >>> list.is_empty() # Returns True/False.
+    True
+    >>> list.insert_at_start(4) # Returns None. params: value. Inserts at beginning of list.
+
+    >>> list.insert_at_end(5) # Returns None. params: value. Insert at end of list.
+
+    >>> list.find(4) # Returns node at which value is present.
+        # Else, raises ValueError
+        # params: value. Node data to search for.
+    [4]
+
+    >>> list.delete(4) # Returns None. params: value. delete element with value.
+    >>> list.insert_at_start(2)
+    >>> list.length() # returns length of linked list
+    2
+    >>> print list # returns string containing elements in list.
+    [2] -> [5]
+    """
+    def __init__(self, head=None):
+        self.head = head
+
+    def is_empty(self):
+        return self.head == None
+
+    def insert_at_start(self, value):
+        temp_node = LinkedListNode(value)
+        temp_node.set_next(self.head)
+        self.head = temp_node
+        return
+
+    def insert_at_end(self, value):
```

Comment:
This method would fail if called on an empty list.
You could fix it by:
```
if self.length() == 0: # or even better len(self). Check next comment.
    return self.insert_at_start(value)
```